### ISSUE TITLE
Add mean_token_accuracy from TRL's `SFTTrainer` to `Trainer`

### ISSUE THREAD

**Issue by RylanSchaeffer** [huggingface/transformers#40303]
### Feature request

TRL's `SFTTrainer` has a metric called `mean_token_accuracy`: https://github.com/huggingface/trl/blob/main/trl/trainer/sft_trainer.py#L1095-L1123:

```

                shift_logits = outputs.logits[..., :-1, :].contiguous()
                shift_labels = inputs["labels"][..., 1:].contiguous()

                # When using Prompt Tuning, skip the virtual tokens in logits before accuracy computation, since they do
                # not correspond to actual input labels.
                shift_logits = shift_logits[:, self.num_virtual_tokens :, :]

                # Get predictions
                predictions = shift_logits.argmax(dim=-1)

                # Create mask for non-padding tokens (assuming ignore_index is -100)
                mask = shift_labels != -100

                # Calculate accuracy only on non-padding tokens
                correct_predictions = (predictions == shift_labels) & mask
                total_tokens = mask.sum()
                correct_tokens = correct_predictions.sum()

                # Gather the correct_tokens and total_tokens across all processes
                correct_tokens = self.accelerator.gather_for_metrics(correct_tokens)
                total_tokens = self.accelerator.gather_for_metrics(total_tokens)

                # Compute the mean token accuracy and log it
                total_sum = total_tokens.sum()
                accuracy = (correct_tokens.sum() / total_sum).item() if total_sum > 0 else 0.0
                self._metrics[mode]["mean_token_accuracy"].append(accuracy)
```

I like this metric and I'd like for this metric to be optionally be added to `transformer`'s `Trainer`.

### Motivation

(Mean) Token Accuracy is a useful additional metric that I'd like the option of additionally logging.

**Comment by Rocketknight1** [comment]
cc @sunmarc @mekkcyber

**Comment by badwriter123** [comment]
Hello @RylanSchaeffer ,

I just went through the TRL's SFTTrainer  mean_token_accuracy metric and there are some places where the metric can be improved. 
For example :
1. We can generalize it, The traditional code assumes only LM setup and it could be made more general to support seq2seq models
2. We can configure the ignore_index, instead of it being hardcoded which causes the code to be changed while using / changing datasets. we can pass it as a parameter and That way, different datasets or tokenizer can adapt without requiring code changes everytime.

what do you think of these improvements ???

**Comment by RylanSchaeffer** [comment]
1. Sure
2. Sure

To be clear, I'm not affiliated with that code. I'm just saying it's useful and I'd like it optionally logged when using `Trainer`

**Comment by qgallouedec** [comment]
Hey @RylanSchaeffer, sorry for the slow reply, catching up on the backlog.

`mean_token_accuracy` is specific to next-token-prediction losses, so it does not generalize across the much wider range of tasks core Trainer serves (classification, regression, seq2seq, object detection, audio, etc), and computing it means keeping full vocab logits around during training, which cuts against ongoing work to avoid materializing them for memory reasons.

TRL's SFTTrainer already has this for the SFT case, and you can get the same behavior in vanilla Trainer today by overriding `compute_loss` in a subclass and calling `self.log(...)` yourself.

Closing as out of scope for core, but feel free to raise further improvements to the metric itself over on TRL.
